### PR TITLE
Added env variable support for himl

### LINKED ISSUE #52: Retrieve values from environment variables
### Expected Behaviour

```
test: 1
home_path: "{{env(HOME)}}"
```

=>

```
test: 1
home_path: /home/zadkiel
```

### Actual Behaviour

```
test: 1
home_path: "{{env(HOME)}}"
```

=>

```
test: 1
```

### Details

Retrieving values from the environment would be a great addition.
Currently it can be done with YQ after the values compilation but it does not support the hierarchical overrides that himl allows.

### PR DESCRIPTION
**Pull request by ecojan**
This PR is meant to bring support to Himl for including environment variables when interpolating.

## Description

Added support for HIML to use env variables when interpolating the hierarchy.

## Related Issue

Resolves #52 

## Motivation and Context

See description. Also can be used to include build-time variables in a HIML structure.

## How Has This Been Tested?

Tested out by adding a field in one of the examples with
```yaml
test: {{env(KUBECONFIG)}}
```

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -267,6 +267,11 @@ remote_states:
 endpoint: "{{outputs.cluster_composition.output.value.redis_endpoint}}"
 ```
 
+### Merge with env variables
+```yaml
+kubeconfig_location: {{env(KUBECONFIG}}
```
Comment (amuraru) [inline README.md:272]:
```suggestion
kubeconfig_location: {{env(KUBECONFIG)}}
```

File: himl/inject_env.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+# the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+# OF ANY KIND, either express or implied. See the License for the specific language
+# governing permissions and limitations under the License.
+
+from os import getenv
+
+
+class EnvVarInjector(object):
+    """
+    Resolve secrets in the form:
```
Comment (amuraru) [inline himl/inject_env.py:16]:
```suggestion
    Resolve variables in the form:
```

File: himl/inject_env.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    def __init__(self):
+        return
+
+    def is_interpolation(self, value):
+        return value.startswith('{{') and value.endswith('}}')
+
+    def inject_env_var(self, line):
+        """
+        Check if value is an interpolation and try to resolve it.
+        Uses a cache, in order to not fetch same secret multiple times.
```
Comment (amuraru) [inline himl/inject_env.py:29]:
```suggestion
```

is there a cache involved?

File: himl/inject_env.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        """
+        if not self.is_interpolation(line):
+            return line
+
+        # remove {{ and }}
+        updated_line = line[2:-2]
```
Comment (amuraru) [inline himl/inject_env.py:35]: What about: key: "prefix-{{env(VAR)}}-suffix" ?

File: himl/inject_env.py (hunk 3 above)
Comment (ecojan) [reply himl/inject_env.py:29]: No there isn't, let me clean this up, elements that remained from the secrets injector that I took inspiration from.

File: himl/inject_env.py (hunk 4 above)
Comment (ecojan) [reply himl/inject_env.py:34]: Do we want to offer support for interpolation with prefix/suffix with environment variables? I was thinking something more similar to the secret resolvers

Comment (amuraru) [reply himl/inject_env.py:34]: i guess it's enough now

**Comment by amuraru** [comment]
also a big gap in him is the lack of unit-tests.
I filed an issue to add a unit test suite for all existing functionality in himl: https://github.com/adobe/himl/issues/56

/cc @danielcoman

File: himl/interpolation.py [hunk 5]
```diff
@@ -83,6 +84,15 @@ def resolve_secrets(self, data, default_aws_profile):
         return data
 
 
+class EnvVarResolver(object):
+    def resolve_env_vars(self, data):
+        # TO DO, ADD Logic for env_vars resolver
```
Comment (amuraru) [inline himl/interpolation.py:89]: what is this TODO referring to?

**Review by amuraru — APPROVED** [review]

File: himl/interpolation.py (hunk 5 above)
Comment (ecojan) [reply himl/interpolation.py]: Removed! Was added at first to track where changes are needed.

**Review by alungu — APPROVED** [review]

File: himl/inject_env.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+        if not self.check_supported_function(updated_line):
+            return line
+
+        # remove env( and ) to extract the env Variable
+        updated_line = updated_line[4:-1]
+
+        # If env variable is missing or not set, the output will be None
+        return getenv(updated_line)
+
+    def check_supported_function(self, value):
```
Comment (costimuraru) [inline himl/inject_env.py:46]: For clarity, how about naming this `is_env_interpolation`?

File: himl/inject_env.py [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+        return value.startswith('env(') and value.endswith(')')
+
+    def split_function_items(self, line):
```
Comment (costimuraru) [inline himl/inject_env.py:49]: Is this used anywhere?

File: himl/inject_env.py (hunk 6 above)
Comment (ecojan) [reply himl/inject_env.py]: Done

File: himl/inject_env.py (hunk 7 above)
Comment (ecojan) [reply himl/inject_env.py]: Good catch, I was using it initially but dropped it and wasn't caught in the cleanup.
I've removed it

**Review by costimuraru — APPROVED** [review]

**Pull request merged**